massa-event-cache/src/controller.rs

```diff
@@ -0,0 +1,117 @@
+// std
+use std::collections::VecDeque;
+use std::sync::Arc;
+// third-party
+use parking_lot::{Condvar, Mutex, RwLock};
+// internal
+use crate::event_cache::EventCache;
+use massa_models::execution::EventFilter;
+use massa_models::output_event::SCOutputEvent;
+
+/// structure used to communicate with controller
+#[derive(Debug)]
+pub(crate) struct EventCacheWriterInputData {
+    /// set stop to true to stop the thread
+    pub stop: bool,
+    pub(crate) events: VecDeque<SCOutputEvent>,
+}
+
+impl EventCacheWriterInputData {
+    pub fn new() -> Self {
+        Self {
+            stop: Default::default(),
+            events: Default::default(),
+        }
+    }
+
+    /// Takes the current input data into a clone that is returned,
+    /// and resets self.
+    pub fn take(&mut self) -> Self {
```

Comment:
if the structure implements Default, you do not need to implement `take` explicitly, it would work with just mem::take(&mut event_cache_writer_input_data)

Reply:
hmm not really

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5e171127d993700e68e42b96a13364a0

Reply:
For the sake of clarity, std::mem::take do require trait Default impl for T as described here: https://doc.rust-lang.org/core/mem/fn.take.html. 

So this will not work:

```
#[derive(Debug, Clone, Default)]
struct Foo2 {
    i: u32,
}

#[derive(Debug, Clone)]
struct Wrapper3<T>(T);

fn main() {
    let f2 = Foo2 { i: 5 };
    let mut w = Wrapper3(f2);
    let f2_new = core::mem::take(&mut w);
}
```

but this will work:

```
#[derive(Debug, Clone, Default)]
struct Foo2 {
    i: u32,
}

#[derive(Debug, Clone)]
struct Wrapper3<T>(T);

impl<T> Deref for Wrapper3<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T> DerefMut for Wrapper3<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

fn main() {
    let f2 = Foo2 { i: 5 };
    let mut w = Wrapper3(f2);
    let f1: Foo2 = core::mem::take(&mut w);
}
```

And this only works because:
* Return type is annotated
* Wrapper3 implements Deref & DerefMut traits (MutexGuard also impl these traits)...

Reply:
Fixed